### PR TITLE
Add context shorthand title top bar and context performance summary bottom bar

### PR DESCRIPTION
**Pull request by mzgoddard**
## Changes

- Add top bar with context shorthand description
- Add bottom bar with real time context performance information
- Fix optional chaining test issue with babel optional chaining transform
- Convert many `src/panel` files into typescript files
- Convert `src/utils/Observer` into a typescript file to assist with ts and rx conversions
- Add `ContextRealtimeData` polling `Observable`
- Add `ContextRealtimeData` member to `Audion.GraphContext`
- Make `AudioGraphRender` more resilient to missing nodes from when a context is created before the panel is opened
- Add devtools frontend css for ui colors to help match look and feel of builtin devtools

-----

### Components

This changeset uses `rxjs` to map input data to text and html that are set to an element provided by a separate observable. Being observables they must be observed to operate, so to operate as an application, the observables are merged into one element observable and subscribed to. A subscriber given to that subscribe call would observe an element for each time an element is updated by this configuration.

```js
// The RealtimeSummary component
export function renderRealtimeSummary(
  element$: Observable<HTMLElement>,
  data$: Observable<Audion.ContextRealtimeData>,
) {
  const realtimeHTML$ = data$.pipe(map(realtimeSummaryHTML));
  return setElementHTML(element$, realtimeHTML$);
}
```

The realtime summary component for example takes a observable of elements and an observable of `ContextRealtimeData` values. The `ContextRealtimeData` value is further formatted as html. Using a helper, `setElementHTML`, from `src/panel/components/domUtils`, the latest element from the element observable has its `innerHTML` assigned to the latest html from the html formatted `ContextRealtimeData` observable.

#### SelectGraph component

As of this changeset the select graph component shows the shorthand title of the current audio context and updates a hidden list of all audio context shorthand titles to be used in a modal dropdown. It approriately updates when a new context is created and the current graph automated selection scheme chooses the newest context. Similarly the hidden list updates with all the shorthand titles.

### CODE REVIEW

File: src/chrome/DebuggerWebAudioDomain.ts [hunk 1]
```diff
@@ -1,27 +1,26 @@
 import {ProtocolMapping} from 'devtools-protocol/types/protocol-mapping';
 
-export const Methods: {
-  [K in string]: keyof ProtocolMapping.Commands & `WebAudio.${K}`;
-} = {
-  disable: 'WebAudio.disable',
-  enable: 'WebAudio.enable',
-  getRealtimeData: 'WebAudio.getRealtimeData',
-};
+export enum WebAudioDebuggerMethod {
```
Comment (hoch) [inline src/chrome/DebuggerWebAudioDomain.ts:3]: When we say "Debugger", what does it exactly mean?

File: src/devtools/WebAudioRealtimeData.ts [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+export const INITIAL_CONTEXT_REALTIME_DATA = {
+  callbackIntervalMean: 0,
+  callbackIntervalVariance: 0,
+  currentTime: 0,
+  renderCapacity: 0,
+} as Audion.ContextRealtimeData;
+
+export class WebAudioRealtimeData {
+  pollContext(contextId: string) {
+    return interval(1000).pipe(
```
Comment (hoch) [inline src/devtools/WebAudioRealtimeData.ts:27]: Can we call "1000" out as a variable?

File: src/panel.html [hunk 3]
```diff
@@ -1,15 +1,264 @@
-<html>
+<html class="-theme-with-dark-background">
   <head>
     <style>
+      body {
```
Comment (hoch) [inline src/panel.html:4]: Just curious - how is this constructed? The style seems very sophisticated.

File: src/utils/toRX.ts [hunk 4]
```diff
@@ -0,0 +1,13 @@
+import {Observable} from 'rxjs';
+
+import {Utils} from './Types';
+
+export function toRX<T>(observer: Utils.Observer<T>): Observable<T> {
```
Comment (hoch) [inline src/utils/toRX.ts:5]: What does this do? Can we have a general comment of the function?

File: src/panel/components/selectGraph.ts [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+  combineLatest,
+  distinctUntilChanged,
+  map,
+  merge,
+  Observable,
+} from 'rxjs';
+import {Audion} from '../../devtools/Types';
+import {setElementHTML, setElementText} from './domUtils';
+
+function graphIdTitle(graphId: string) {
```
Comment (hoch) [inline src/panel/components/selectGraph.ts:11]: It would be great if we can have general comments on each function.

**Review by hoch — COMMENTED** [review]
Although I am an expert of TS, the code base now looks a bit simpler. It looks general great, but a general comment on functions would be helpful for the future maintenance.

**Comment by hoch** [comment]
I also noticed there is an empty file for the future usage. I think it's best to remove them if it is not required by the current revision/commit.

File: src/chrome/DebuggerWebAudioDomain.ts (hunk 1 above)
Comment (mzgoddard) [reply src/chrome/DebuggerWebAudioDomain.ts:3]: Its a value passed to the [`chrome.debugger`](https://developer.chrome.com/docs/extensions/reference/debugger/#method-sendCommand) api and its defined by [chrome devtools protocol web audio domain](https://chromedevtools.github.io/devtools-protocol/tot/WebAudio/). So we could call it `WebAudioDevtoolsProtocolMethod` alternatively.

Comment (hoch) [reply src/chrome/DebuggerWebAudioDomain.ts:3]: Thanks for the explanation!

File: src/panel.html (hunk 3 above)
Comment (mzgoddard) [reply src/panel.html:4]: These are excerpts from the following:

- https://github.com/ChromeDevTools/devtools-frontend/blob/main/front_end/ui/legacy/inspectorCommon.css#L81-L150
- https://github.com/ChromeDevTools/devtools-frontend/blob/main/front_end/ui/legacy/themeColors.css#L13-L336

I've been thinking about how to manage these. (Probably namely comments referring to those sources, but also to maybe put it in their own files and either `<link>` them or inject them in a build step.) I also want to pair them down more. I removed some colors already, and want to remove most of the rest. I think pair down the fonts or remove them altogether and use `sans-serif` alone.

Comment (mzgoddard) [reply src/panel.html:4]: Since I've merged this change I made an issue to help track this https://github.com/GoogleChrome/audion/issues/126.

**Comment by hoch** [comment]
I am approving the PR after the merge. :)

**Pull request merged**